Additional refactoring of Null Semantics

maumar opened the pull request:
- moving NullSemantics visitor after 2nd level cache - we need to know the parameter values to properly handle IN expressions wrt null semantics,
- NullSemantics visitor needs to go before SqlExpressionOptimizer and SearchCondition, so those two are also moved after 2nd level cache,
- moving optimizations that depend on knowing the nullability to NullSemantics visitor - optimizer now only contains optimizations that also work in 3-value logic, or when we know nulls can't happen,
- merging InExpressionValuesExpandingExpressionVisitor int NullSemantics visitor, so that we don't apply the rewrite for UseRelationalNulls,
- preventing NulSemantics from performing double visitation when computing non-nullable columns.

Resolves #11464
Resolves #15722
Resolves #18338
Resolves #18597
Resolves #18689
Resolves #19019

maumar commented:
@roji

maumar commented on src/EFCore.Relational/Query/Internal/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -148,7 +148,11 @@ protected virtual Expression VisitSqlUnaryExpression(SqlUnaryExpression sqlUnary
+                        // NullSemantics removes all nulls from expressions wrapped around Not
+                        // so the optimizations are safe to do as long as UseRelationalNulls = false
+                        case SqlBinaryExpression binaryOperand
+                            when !_useRelationalNulls:
```
shall we do those optimizations always (i.e. even when relational nulls == true)? We do that currently which is technically wrong, but perhaps for relational nulls all bets are off

roji replied:
At the very least, some stuff here looks like it's correct even for 3-value logic (e.g. `!(a == b) -> (a != b)`), no?

In any case, if something is incorrect for 3-value logic I guess we shouldn't do it, right?

maumar replied:
sounds good!

maumar commented on src/EFCore.Relational/Query/RelationalParameterBasedQueryTranslationPostprocessor.cs:
```diff
@@ -38,17 +38,19 @@ public class RelationalParameterBasedQueryTranslationPostprocessor
                 canCache = false;
             }
 
             var nullParametersOptimized = new ParameterNullabilityBasedSqlExpressionOptimizingExpressionVisitor(
```
ParameterNullabilityBasedSqlExpressionOptimizingExpressionVisitor - name of that visitor is a bit redundant now, since it always knows the parameter values. Are we ok with breaking providers and fixing the names?

maumar replied:
TODO: will clean this up: move _parameterValues to the base visitor and remove this one.

maumar commented on test/EFCore.SqlServer.FunctionalTests/Query/NorthwindMiscellaneousQuerySqlServerTest.cs:
```diff
@@ -343,7 +343,7 @@ public override async Task Default_if_empty_top_level(bool async)
     FROM [Employees] AS [e]
     WHERE [e].[EmployeeID] = -1
-) AS [t] ON 1 = 1");
+) AS [t] ON CAST(1 AS bit) = CAST(1 AS bit)");
```
for sql server, should we optimize `CAST(1 AS bit) = CAST(1 AS bit)` to `1 = 1` and `CAST(0 AS bit) = CAST(1 AS bit)` to `0 = 1` everywhere?

roji replied:
Sure, SQL would definitely be nicer!

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -111,56 +123,163 @@ protected override Expression VisitIn(InExpression inExpression)
+            {
+                case SqlConstantExpression sqlConstant:
+                {
+                    CanCache = false;
```
disable caching when dealing with constant array, or just parameter?

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -370,169 +510,309 @@ protected override Expression VisitSqlBinary(SqlBinaryExpression sqlBinaryExpres
-                            _sqlExpressionFactory.Equal(newLeft, newRight),
-                            _sqlExpressionFactory.AndAlso(leftIsNull, rightIsNull));
-                    }
+                // TODO: should we do this for parameters also?
```
if we do, we will prune some of the parameters, converting them to constants effectively. It shouldn't explode our cache since its only for bool values, so perhaps it's worth doing. Thoughts?

maumar commented on src/EFCore.SqlServer/Query/Internal/SqlServerParameterBasedQueryTranslationPostprocessor.cs:
```diff
@@ -25,7 +26,10 @@ public class SqlServerParameterBasedQueryTranslationPostprocessor : RelationalPa
                 .Visit(optimizedSelectExpression);
 
-            return (searchConditionOptimized, canCache);
+            var optimized = (SelectExpression)new SqlExpressionOptimizingExpressionVisitor(
```
we need to run this second time again (before those optimizations were made by 2nd level cache optimizer, where SearchCondition was done earlier).

maumar commented on src/EFCore.SqlServer/Query/Internal/SqlServerQueryTranslationPostprocessor.cs:
```diff
@@ -15,13 +14,5 @@ public class SqlServerQueryTranslationPostprocessor : RelationalQueryTranslation
             : base(dependencies, relationalDependencies, queryCompilationContext)
         {
         }
-
```
TODO: will remove (or shall we keep the scaffolding? ;) )

smitpatel replied:
remove internal?
@ajcvickers

---
In reply to: [354103808](https://github.com/aspnet/EntityFrameworkCore/pull/19168#discussion_r354103808) [](ancestors = 354103808)

maumar commented on src/EFCore.Relational/Query/RelationalParameterBasedQueryTranslationPostprocessor.cs:
```diff
@@ -38,17 +37,19 @@ public class RelationalParameterBasedQueryTranslationPostprocessor
 
-            var nullParametersOptimized = new ParameterNullabilityBasedSqlExpressionOptimizingExpressionVisitor(
-                Dependencies.SqlExpressionFactory, UseRelationalNulls, parametersValues).Visit(inExpressionOptimized);
+            var nullParametersOptimized = new SqlExpressionOptimizingExpressionVisitor(
```
should we create those using DI, or should it be enough for providers to just override Optimize method?

maumar replied:
or perhaps we can add optimize method so that providers can inject their optimizations after our visitor runs, but before FromSqlParameterApplyingExpressionVisitor

maumar replied:
https://github.com/aspnet/EntityFrameworkCore/issues/10514

maumar replied:
will be done in separate PR, if we decide to do it

roji commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -370,169 +518,311 @@ protected override Expression VisitSqlBinary(SqlBinaryExpression sqlBinaryExpres
+                // !a == b <=> a == !b -> a != b
+                // a != b <=> !a != !b -> a != b
+                // !a != b <=> a != !b -> a == b
+                return sqlBinaryExpression.OperatorType == ExpressionType.Equal
```
Can simplify with XOR?

roji commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -567,31 +857,232 @@ protected override Expression VisitSqlFunction(SqlFunctionExpression sqlFunction
         {
-            // at this point we assume every parameter is nullable, we will filter out the non-nullable ones once we know the actual values
-            _isNullable = true;
+            _isNullable = _parameterValues[sqlParameterExpression.Name] == null;
```
I wonder... we treat null parameters as nullable, but if we treated them as a null *constant* instead, some other optimizations could open up. How about returning an SqlConstantExpression here?

maumar replied:
Sounds good

roji commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -567,31 +857,232 @@ protected override Expression VisitSqlFunction(SqlFunctionExpression sqlFunction
+                : result;
+        }
+
+        private SqlExpression OptimizeNonNullableNotExpression(SqlUnaryExpression sqlUnaryExpression)
```
Should this also be called after rewriting null semantics?

maumar replied:
i think it's not needed - null semantics rewrite never ends up with top level IsNull/IsNotNull expression, so this optimization would always be a noop. We run the optimizations on IsNull parts of the null semantics rewrite, so they should be in their minimal form already.

roji commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -567,31 +857,232 @@ protected override Expression VisitSqlFunction(SqlFunctionExpression sqlFunction
+        private SqlExpression OptimizeNonNullableNotExpression(SqlUnaryExpression sqlUnaryExpression)
+        {
+            // optimizations below are only correct in 2-value logic
+            if (sqlUnaryExpression.Operand is SqlBinaryExpression sqlBinaryOperand)
```
Since the whole method is only interested in binary, I always prefer to inverse the logic and exit early for non-binary, this way we reduce an indentation (easier to read).

roji commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -567,31 +857,232 @@ protected override Expression VisitSqlFunction(SqlFunctionExpression sqlFunction
+                        sqlBinaryOperand.TypeMapping);
+                }
+
+                static bool TryNegate(ExpressionType expressionType, out ExpressionType result)
```
Probably better to put local functions at the very end.

maumar replied:
I thought we were supposed to put local functions at the end of relevant block. I was wrong ;) - fixed

roji commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -567,31 +857,232 @@ protected override Expression VisitSqlFunction(SqlFunctionExpression sqlFunction
+                    // (a ?? b) == null -> a == null && b == null
+                    // (a ?? b) != null -> a != null || b != null
+                    // for AndAlso, OrElse we can't do this optimization
+                    // we could do something like this, but it seems too complicated:
```
Out of curiosity, do we think it's too complicated for us to do, or that it would degrade the overall perf of the SQL query?

maumar replied:
the latter - SQL becomes way too complicated

roji commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -2,25 +2,37 @@
 
 namespace Microsoft.EntityFrameworkCore.Query.Internal
 {
     public class NullSemanticsRewritingExpressionVisitor : SqlExpressionVisitor
```
We should consider renaming this as it now also does general null optimizations (maybe NullabilityHandlingExpressionVisitor or similar?)

smitpatel commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -111,56 +123,164 @@ protected override Expression VisitIn(InExpression inExpression)
+                    var values = (IEnumerable)sqlConstant.Value;
+                    foreach (var value in values)
+                    {
+                        if (value == null)
```
We don't translate above yet. It requires representation of array on server side.

roji replied:
Am I missing something? Can't the above simply be translated as:

```c#
SELECT *
FROM Customers
WHERE 2 IN (1, Id, 3)
```

smitpatel replied:
It can be. We did not have customer feedback to add it yet and it works only for Contains.

roji replied:
OK. Seems pretty simple to me, but we can forget for now if people don't think it's important.

smitpatel commented:
>  we can also carry info from right to left just as well.

That may not be correct. The logic of carrying over non-nullability of left to right relies on short-circuiting behavior of `&&` operator. logical operators does not short-circuit from right side.

roji commented:
>> we can also carry info from right to left just as well.
>
> That may not be correct. The logic of carrying over non-nullability of left to right relies on short-circuiting behavior of && operator. logical operators does not short-circuit from right side.

From my understanding SQL doesn't guarantee any particular evaluation order: [PG docs on this](https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-EXPRESS-EVAL), [some](https://blogs.msdn.microsoft.com/bartd/2011/03/03/dont-depend-on-expression-short-circuiting-in-t-sql-not-even-with-case/) [SQL Server](https://www.sqlservercentral.com/articles/understanding-t-sql-expression-short-circuiting) [links](https://weblogs.sqlteam.com/jeffs/2008/02/22/sql-server-short-circuit/) (I think we discussed this briefly in Redmond). Unless I'm mistaken the query planner is free to evaluate the right-hand side first if it thinks it's cheaper etc.

Also, I'm not 100% sure, but I'm not sure this matters for conjunctions. If one side requires that a column be non-null, then the conjunction can only be true if that is the case, so it shouldn't matter whether the other side contains null-compensating constructs or not... And so we should be able to omit them.

smitpatel commented:
While, I was not aware of SQL doing differently, what I wrote pertains to C# short-circuiting behavior. (`&&` operator and not `AND`). The likely way to have short-circuiting arising in SQL tree is going to be from C# behavior (and in most cases user written). So we are capturing most cases where it could be useful right now.

We could check right side also and try to apply left side though first thing would be we need to evaluate if it is going to be correct in all cases. Next, we need to evaluate if cost of doing such processing is worth the perf penalty when C# code would never generate it (hence falls down to translation pipeline generate such code) and translation pipeline can translate it in C# way of short-circuiting. Worst case, it won't be optimized and would be longer SQL.

Even if it is correct, given perf cost, we should not implement it unless we have a specific user reported case where it would be required and it is not arising before issue in pipeline somewhere.

Feel free to file an issue if you feel strongly about it.

roji commented:
@smitpatel as long as we're implementing this new mechanism (nonNullableColumns), I don't see why we shouldn't just do it right off the bat... we implement many things not explicitly requested by users because we know they're right.

I agree that user-written code is more likely to assume left-to-right short-circuiting behavior (as in C#). However, we also apply various transformations that could result in right-to-left optimization opportunities, making this worthwhile IMHO. For example, consider something like this:

```c#
customers.Where(c => c.Name != x && c.Name > 3)
```

This seems like pretty natural C# code. However, the right-hand side can only be true if c.Name is non-nullable; this could be carried over to the left side, obviating null compensation there.

Maybe before deferring this to another issue we should see what @maumar thinks?

smitpatel commented:
nonNullable columns is not new thing, it was already there before too.

> Feel free to file an issue if you feel strongly about it.

roji commented:
Opened #19248

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -111,56 +123,164 @@ protected override Expression VisitIn(InExpression inExpression)
+                        : _sqlExpressionFactory.IsNull(item);
+            }
+
+            _isNullable = _isNullable || itemNullable || subqueryNullable;
```
need to set _isNullable before returning from the function below - simplified and moved the logic inside the if block

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -111,56 +123,164 @@ protected override Expression VisitIn(InExpression inExpression)
+                // non_nullable NOT IN (1, 2) -> non_nullable NOT IN (1, 2)
+                // non_nullable NOT IN (1, 2, NULL) -> non_nullable NOT IN (1, 2)
+                // nullable IN (1, 2) -> nullable IN (1, 2) (optimized)
+                return inExpression.Update(item, inValues, subquery);
```
there could have been null in the values, example 2 and 4 in the comment

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -111,56 +123,164 @@ protected override Expression VisitIn(InExpression inExpression)
-            var newEscapeChar = (SqlExpression)Visit(likeExpression.EscapeChar);
-            _isNullable |= isNullable;
+            var newTable = (TableExpressionBase)Visit(innerJoinExpression.Table);
+            var newJoinPredicate = VisitJoinPredicate((SqlBinaryExpression)innerJoinExpression.JoinPredicate);
```
c# join follows relational null semantics - we do this to explicitly avoid null semantics rewrites. However, to further complicate things ,if you compare composite keys the comparison behaves like any other c# comparison.
So: `join null equals null` wouldn't match
but: `join { null } equals { null }` would

roji replied:
Oh, this is new to me...! Very strange that C# has relational null semantics in this one, unique place. So this makes sense...

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -181,11 +301,24 @@ private SqlExpression VisitJoinPredicate(SqlBinaryExpression predicate)
+                var leftNullable = _isNullable;
+                _isNullable = false;
+                var right = (SqlExpression)Visit(predicate.Right);
+                var rightNullable = _isNullable;
```
it doesn't matter - VisitJoinPredicate is always called from JOIN, which then again is part of SelectExpression which is always nullable

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -370,169 +518,311 @@ protected override Expression VisitSqlBinary(SqlBinaryExpression sqlBinaryExpres
-                    && !rightNegated)
+            bool? IsTrueOrFalse(SqlExpression sqlExpression)
+            {
+                if (sqlExpression is SqlConstantExpression sqlConstantExpression && sqlConstantExpression.Value is bool boolConstant)
```
yeah, i realized that as well

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -111,56 +123,164 @@ protected override Expression VisitIn(InExpression inExpression)
+            // nullable NOT IN (1, 2, NULL) -> nullable NOT IN (1, 2) AND nullable IS NOT NULL (full)
+            return inExpression.IsNegated == hasNullValue
+                ? _sqlExpressionFactory.AndAlso(
+                    inExpression.Update(item, inValues, subquery),
```
I implemented it initially, but the more I look at it, the less I like it. Feels like opening a can of worms - we have other functions where not all arguments are always used (CaseExpression). Also, select expression has a ton of arguments and it could use "targeted" update methods. I'd rather keep 1 method which always takes all the arguments and don't worry/think about which functions need what. Thoughts? @roji @smitpatel

roji replied:
As a general rule, if the invariant of the expression is for exactly one of subquery and values to be null, then ideally it shouldn't be possible to create a violating instance. #19229 was about doing that check in runtime, but doing that at compile-time (via the methods we expose is even better). This both creates a safer type, and documents the invariants very nicely.

Also, as discussed in design, we don't have to generalize every such decision - it seems OK to do this for InExpression, and if we feel that SelectExpression is a different case, have something different. But let's see what @smitpatel has to say :)

smitpatel replied:
Single method.

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -567,31 +857,232 @@ protected override Expression VisitSqlFunction(SqlFunctionExpression sqlFunction
+
+        private SqlExpression ProcessNullNotNull(
+            SqlUnaryExpression sqlUnaryExpression,
+            SqlExpression operand,
```
fixed case where it was different and removed the argument

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -567,31 +857,232 @@ protected override Expression VisitSqlFunction(SqlFunctionExpression sqlFunction
+                    return sqlConstantOperand.Value == null
+                        ? _sqlExpressionFactory.Constant(
+                            sqlUnaryExpression.OperatorType == ExpressionType.Equal,
+                            sqlUnaryExpression.TypeMapping)
```
don't have access to it

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -111,56 +123,164 @@ protected override Expression VisitIn(InExpression inExpression)
+
+            var processedValues = inValues.Count > 0
+                ? (SqlExpression)Visit(_sqlExpressionFactory.Constant(inValues, typeMapping))
+                : null;
```
makes the check for empty more complicated - need to cast to SqlConstantExpression (or we can assume result is always SqlConstant, but providers could change that behavior), then cast result to List<object> and then test Count. I can do that but not sure if it's much better.

maumar replied:
actually, after giving it some more thought I agree that empty list looks better overall - will change

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -111,56 +123,164 @@ protected override Expression VisitIn(InExpression inExpression)
+            }
+
+            var processedValues = inValues.Count > 0
+                ? (SqlExpression)Visit(_sqlExpressionFactory.Constant(inValues, typeMapping))
```
added a comment to fix this when/if we support mix of columns and constants

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -370,169 +518,311 @@ protected override Expression VisitSqlBinary(SqlBinaryExpression sqlBinaryExpres
+
+            // a == null -> a IS NULL
+            // a != null -> a IS NOT NULL
+            if (rightNullValue)
```
I'd rather keep the logic as simple as possible, even when it means some duplication.

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -370,169 +518,311 @@ protected override Expression VisitSqlBinary(SqlBinaryExpression sqlBinaryExpres
+            // only correct in 2-value logic
+            // a == a -> true
+            // a != a -> false
+            if (!leftNullable
```
Current optimizations shouldn't make a difference - all cases where rewrite null semantics does something (i.e. weeds out nulls from the expression) and up with AndAlso or OrElse as a top level node, whereas all the optimizations here are for == and !=. Might be worthwhile in the future if/when we come up with more optimizations

maumar commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -181,11 +301,24 @@ private SqlExpression VisitJoinPredicate(SqlBinaryExpression predicate)
+                var right = (SqlExpression)Visit(predicate.Right);
+                var rightNullable = _isNullable;
+
+                var result = OptimizeComparison(
```
Not sure what you mean. It should only end up as cross join if the join predicate collapses to tautology, which seems fine to me. And we run those optimizations for inner and outer joins both.

roji replied:
Yeah, that's pretty much what I meant. If OptimizeComparison reduces the expression to true, we have a cross join. The question is if in Visit{Inner,Left}Join we should check for this case and return a CrossJoinExpression, instead of having an inner/left join with a TRUE constant predicate.

maumar replied:
good idea - done

maumar commented:
@roji @smitpatel new version is up, please take a look. Biggest concern is NullHandling, parts that compute non-nullable columns (especially in binary, but also I might have missed a reset somewhere >.<)

roji commented on src/EFCore.Relational/Query/Internal/NullabilityHandlingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1474 @@
+            var hasNullValue = false;
+            RelationalTypeMapping typeMapping = null;
+
+            if (valuesExpression is SqlConstantExpression
```
Just for correctness, what happens if valuesExpression is neither constant nor parameter? I know non-PG database don't support array columns, and even on PG translation will have already transformed the whole thing to something else. But maybe an array with a value converter?

Maybe it's better to just lift this check to VisitIn and just immediately return from there.

maumar commented:
another version up @roji @smitpatel

maumar commented on src/EFCore.Relational/Query/Internal/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -52,12 +38,27 @@ protected override Expression VisitExtension(Expression extensionExpression)
+            {
+                var result = base.VisitExtension(innerJoinExpression);
+
+                return (result is InnerJoinExpression innerJoinExpressionResult
```
we can only optimize INNER JOIN - if we try this with left join, we get incorrect results if the right side has no elements

smitpatel commented on src/EFCore.Relational/Query/RelationalParameterBasedQueryTranslationPostprocessor.cs:
```diff
@@ -46,17 +45,19 @@ public class RelationalParameterBasedQueryTranslationPostprocessor
-                Dependencies.SqlExpressionFactory, parametersValues).Visit(selectExpression);
-
-            if (!ReferenceEquals(selectExpression, inExpressionOptimized))
+            var nullabilityHandled = nullabilityHandlingExpressionVisitor.Visit(selectExpression);
```
Use same tuple pattern for return type.

smitpatel commented on src/EFCore.Relational/Query/Internal/NullabilityHandlingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1472 @@
+
+        private void RestoreNonNullableColumnsList(int counter)
+        {
+            _nonNullableColumns.RemoveRange(counter, _nonNullableColumns.Count - counter);
```
Add a check before calling into RemoveRange if there are additional columns

smitpatel commented on src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs:
```diff
@@ -567,31 +857,232 @@ protected override Expression VisitSqlFunction(SqlFunctionExpression sqlFunction
+            return sqlUnaryExpression;
+        }
+
+        private SqlExpression ProcessNullNotNull(
```
Why is this protected virtual, who is overriding it and where?

maumar commented:
@smitpatel @roji new version up, addressing feedback from design meeting, mostly combining null semantics and sql optimizer and applying some optimizations in search condition, so that we don't need to run optimizer after search condition

smitpatel commented on src/EFCore.Relational/Query/FromSqlParameterApplyingExpressionVisitor.cs:
```diff
@@ -0,0 +1,134 @@
+                    switch (fromSql.Arguments)
+                    {
+                        case ParameterExpression parameterExpression:
+                            var parameterValues = (object[])_parametersValues[parameterExpression.Name];
```
Could DRY

maumar replied:
too many differences

smitpatel commented:
>         public virtual (SelectExpression selectExpression, bool canCache) Optimize(

Tuple members are PascalCase, when you assign them to variable they are camelCase. @AndriySvyryd - thoughts?

---
Refers to: src/EFCore.Relational/Query/RelationalParameterBasedQueryTranslationPostprocessor.cs:41 in 5c11423. [](commit_id = 5c114233ab3a9982823640895a2b900afbb6d897, deletion_comment = True)

smitpatel commented on src/EFCore.Relational/Query/RelationalParameterBasedQueryTranslationPostprocessor.cs:
```diff
@@ -46,300 +32,24 @@ public class RelationalParameterBasedQueryTranslationPostprocessor
                 canCache = false;
             }
 
             return (selectExpression: (SelectExpression)fromSqlParameterOptimized, canCache);
```
Avoid named parameter in this case

smitpatel commented on src/EFCore.Relational/Query/RelationalQueryTranslationPostprocessor.cs:
```diff
@@ -42,22 +38,12 @@ public override Expression Process(Expression query)
             return query;
         }
 
         protected virtual Expression OptimizeSqlExpression([NotNull] Expression query)
```
Obsolete if not breaking
@ajcvickers

maumar replied:
seems like it could be useful for some providers wanting to add optimizations that can be done before 2nd level cache

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+        protected virtual bool CanCache { get; set; }
+
+        public SqlExpressionOptimizingExpressionVisitor(
+            bool useRelationalNulls,
```
bool at the end may be

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+        {
+            Check.NotNull(selectExpression, nameof(selectExpression));
+
+            var result = (SelectExpression)Visit(selectExpression);
```
Inline

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            CanOptimize = false;
+            var newOperand = (SqlExpression)Visit(caseExpression.Operand);
+
+            RestoreNonNullableColumnsList(currentNonNullableColumnsCount);
```
Why operand need this restoration? It should be value type on top level so it would never add itself. If it is nested which contain condition then nested element will remove it.

maumar replied:
i added the restore everywhere, just in case - will remove

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            var isNullable = caseExpression.ElseResult == null;
+
+            var currentNonNullableColumnsCount = NonNullableColumns.Count;
+            var canOptimize = CanOptimize;
```
Keep isnullable and CanOptimize together with their counterparts

maumar replied:
largely obsoleted by using a helper method

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            var newWhenClauses = new List<CaseWhenClause>();
+            foreach (var whenClause in caseExpression.WhenClauses)
+            {
+                CanOptimize = testIsCondition;
```
Reset IsNullable before every visit. It may be redundant in some cases but can become error prone pretty quick

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                isNullable |= IsNullable;
+                newWhenClauses.Add(new CaseWhenClause(newTest, newResult));
+
+                RestoreNonNullableColumnsList(currentNonNullableColumnsCount);
```
This shouldn't be here

maumar replied:
there could be null columns set in the result - we don't want that information to propagate to another WhenClause

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                RestoreNonNullableColumnsList(currentNonNullableColumnsCount);
+            }
+
+            CanOptimize = false;
```
reset

maumar replied:
if null column information is added when visiting else it shouldn't propagate outside the CaseExpression

maumar replied:
added helper method

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            IsNullable |= isNullable;
+            CanOptimize = canOptimize;
+
+            RestoreNonNullableColumnsList(currentNonNullableColumnsCount);
```
why

maumar replied:
if null column information is added when visiting else it shouldn't propagate outside the CaseExpression

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+        {
+            Check.NotNull(columnExpression, nameof(columnExpression));
+
+            IsNullable = !NonNullableColumns.Contains(columnExpression) && columnExpression.IsNullable;
```
put IsNullable check first. Contains on a list is more expensive

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+        {
+            Check.NotNull(crossApplyExpression, nameof(crossApplyExpression));
+
+            var currentNonNullableColumnsCount = NonNullableColumns.Count;
```
DRY with other table expression except for update part.
Optionally can also add abstract method Update on TableExpressionBase

maumar replied:
no need anymore, collapsed to one line

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            CanOptimize = false;
+            var source1 = (SelectExpression)Visit(exceptExpression.Source1);
+            RestoreNonNullableColumnsList(currentNonNullableColumnsCount);
+            var source2 = (SelectExpression)Visit(exceptExpression.Source2);
```
reset canOptimize?

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+
+            var canOptimize = CanOptimize;
+            CanOptimize = false;
+            IsNullable = false;
```
no remembering of previous IsNullable?

maumar replied:
we never do - IsNullable always resets at the start of the visit and we only look at it's value after visitation is complete

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                    inExpression.Update(item, inValues, subquery: null),
+                    SqlExpressionFactory.IsNull(item));
+
+            (SqlConstantExpression processedValues, bool hasNullValue) ProcessInExpressionValues(SqlExpression valuesExpression)
```
Casing

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                    values = (IEnumerable)sqlConstant.Value;
+                }
+
+                if (valuesExpression is SqlParameterExpression sqlParameter)
```
else if

smitpatel replied:
switch block?

---
In reply to: [362989908](https://github.com/aspnet/EntityFrameworkCore/pull/19168#discussion_r362989908) [](ancestors = 362989908)

smitpatel replied:
switch block would also give chance to throw for unknown type which are ignored here.

---
In reply to: [362989948](https://github.com/aspnet/EntityFrameworkCore/pull/19168#discussion_r362989948) [](ancestors = 362989948,362989908)

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            CanOptimize = canOptimize;
+
+            // either values array is empty or only contains null
+            if (((List<object>)inValues.Value).Count == 0)
```
just project out inner list too rather than accessing casting like this

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                    inValues.Add(value);
+                }
+
+                // this is only correct if constant values are the only things allowed here, i.e no mixing of constants and columns
```
Remove comment. Mixing cannot be represented in current SQL tree anyway.

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            CanOptimize = false;
+            var newTable = (TableExpressionBase)Visit(innerJoinExpression.Table);
+            RestoreNonNullableColumnsList(currentNonNullableColumnsCount);
+            var newJoinPredicate = VisitJoinPredicate((SqlBinaryExpression)innerJoinExpression.JoinPredicate);
```
reset canoptimize

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+
+        protected override Expression VisitIntersect(IntersectExpression intersectExpression)
+        {
+            Check.NotNull(intersectExpression, nameof(intersectExpression));
```
DRY all set operations

maumar replied:
with helper method those got reduced to 2 lines - not worth DRYing

smitpatel replied:
Elaborate more. They do exactly same processing. Add abstract Update method in SetOperationBase.

maumar replied:
this is how they look right now:

```cs
            var source1 = (SelectExpression)VisitInternal(intersectExpression.Source1);
            var source2 = (SelectExpression)VisitInternal(intersectExpression.Source2);

            return intersectExpression.Update(source1, source2);
```

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            CanOptimize = false;
+            var source1 = (SelectExpression)Visit(intersectExpression.Source1);
+            RestoreNonNullableColumnsList(currentNonNullableColumnsCount);
+            var source2 = (SelectExpression)Visit(intersectExpression.Source2);
```
reset canOptimize

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            CanOptimize = canOptimize;
+            RestoreNonNullableColumnsList(currentNonNullableColumnsCount);
+
+            return leftJoinExpression.Update(newTable, newJoinPredicate);
```
DRY partial

maumar replied:
non needed due to helper method

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            Check.NotNull(inExpression, nameof(inExpression));
+
+            var canOptimize = CanOptimize;
+            CanOptimize = false;
```
Extract a method out which take expression & canoptimize and returns visited expression and nullability

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                return result;
+            }
+
+            if (predicate.OperatorType == ExpressionType.AndAlso)
```
switch block

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                var newOrdering = (OrderingExpression)Visit(ordering);
+                changed |= newOrdering != ordering;
+                orderings.Add(newOrdering);
+                RestoreNonNullableColumnsList(currentNonNullableColumnsCount);
```
avoid in both foreach

maumar replied:
we need to reset after every visit - we dont want information to propagate from one element to another. Now done in a helper method so its easier on the eyes.

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                projections.Add(updatedProjection);
+                changed |= updatedProjection != item;
+
+                RestoreNonNullableColumnsList(currentNonNullableColumnsCount);
```
Avoid

maumar replied:
we need it so that nullable comuln information doesn't propagate between projections. However now we handle it in the helper method so the code looks cleaner

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+
+            CanOptimize = canOptimize;
+
+            // we assume SelectExpression can always be null
```
It is not our assumption, it is reality lol

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            IsNullable = false;
+            var canOptimize = CanOptimize;
+
+            CanOptimize = CanOptimize && (sqlBinaryExpression.OperatorType == ExpressionType.AndAlso
```
formatting

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            var rightNullable = IsNullable;
+            if (sqlBinaryExpression.OperatorType == ExpressionType.OrElse)
+            {
+                var intersect = leftNonNullableColumns.Intersect(NonNullableColumns).ToList();
```
only compute for the columns which were added in this method rather than all previous ones

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                    left = AddNullConcatenationProtection(left, sqlBinaryExpression.TypeMapping);
+                }
+
+                if (rightNullable)
```
if both nullable only do it once

maumar replied:
need to do it on both sides at least in the case of column

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                // It should be fine because current optimizations *radically* change the expression
+                // (e.g. binary -> unary, or binary -> constant)
+                // but we need to pay attention in the future if we introduce more subtle transformations here
+                if (optimized == updated
```
do reference equality of expecting it be same (no rewrite)

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            SqlExpression AddNullConcatenationProtection(SqlExpression argument, RelationalTypeMapping typeMapping)
+                => argument switch
+                {
+                    SqlConstantExpression _ => SqlExpressionFactory.Constant(string.Empty, typeMapping),
```
cache

maumar replied:
moved from switch expression to elvis

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                    SqlConstantExpression _ => SqlExpressionFactory.Constant(string.Empty, typeMapping),
+                    SqlParameterExpression _ => SqlExpressionFactory.Constant(string.Empty, typeMapping),
+                    ColumnExpression _ => SqlExpressionFactory.Coalesce(argument, SqlExpressionFactory.Constant(string.Empty, typeMapping)),
+                    _ => argument
```
this could give incorrect results

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+
+            bool? IsTrueOrFalse(SqlExpression sqlExpression)
+            {
+                if (sqlExpression is SqlConstantExpression sqlConstantExpression && sqlConstantExpression.Value is bool boolConstant)
```
why only constant and not parameter?

maumar replied:
we dont want to lose parameterized queries here

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                    // not_null_value_parameter is null -> false
+                    // not_null_value_parameter is not null -> true
+                    return SqlExpressionFactory.Constant(
+                        ParameterValues[sqlParameterOperand.Name] == null ^ sqlUnaryExpression.OperatorType == ExpressionType.NotEqual,
```
DRY with constant case

maumar replied:
not worth - abstracting out whether value of Constant/Parameter is null is too much work

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                            // op(a) is not null -> a is not null
+                            return ProcessNullNotNull(
+                                SqlExpressionFactory.MakeUnary(
+                                    sqlUnaryExpression.OperatorType,
```
>sqlUnaryExpression [](start = 36, length = 18)

Could do sqlUnaryExpression.Update(sqlUnaryOperand.Operand)

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                break;
+            }
+
+            //var sqlUnaryOperand = sqlUnaryExpression.Operand as SqlUnaryExpression;
```
?

maumar replied:
removed

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+                        : ExpandNegatedNullableNotEqualNullable(left, right, leftIsNull, leftIsNotNull, rightIsNull, rightIsNotNull);
+                }
+
+                if (leftNullable)
```
use same set of if conditions.

smitpatel replied:
as above block

---
In reply to: [363009379](https://github.com/aspnet/EntityFrameworkCore/pull/19168#discussion_r363009379) [](ancestors = 363009379)

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+
+            var newInstance = (SqlExpression)Visit(sqlFunctionExpression.Instance);
+            RestoreNonNullableColumnsList(currentNonNullableColumnsCount);
+            var newArguments = new SqlExpression[sqlFunctionExpression.Arguments.Count];
```
use niladic flag to avoid allocation

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+            return unionExpression.Update(source1, source2);
+        }
+
+        private List<ColumnExpression> FindNonNullableColumns(SqlExpression sqlExpression)
```
Where is this being used?

maumar replied:
nowhere - removed

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+        // 0 | 0 | 1           | 1                             | 1             |
+        // 0 | 1 | 0           | 1                             | 0             |
+        // 0 | N | N           | 0                             | 0             |
+        // 1 | 0 | 0           | 1            
```
could avoid line breaks if not necessary

maumar replied:
too many arguments, @AndriySvyryd would be unhappy ;)

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1723 @@
+        // 0 | 0 | 1           | 1                             | 1             |
+        // 0 | 1 | 0           | 1                             | 0             |
+        // 0 | N | N           | 0                             | 0             |
+        // 1 | 0 | 0           | 1            
```
?

maumar replied:
removed

smitpatel commented on src/EFCore.SqlServer/Query/Internal/SearchConditionConvertingExpressionVisitor.cs:
```diff
@@ -23,34 +23,58 @@ public class SearchConditionConvertingExpressionVisitor : SqlExpressionVisitor
+                        : _sqlExpressionFactory.Constant(0),
+                    _sqlExpressionFactory.Constant(1))
+                : _sqlExpressionFactory.Equal(
+                sqlExpression,
```
indent

smitpatel commented on src/EFCore.SqlServer/Query/Internal/SearchConditionConvertingExpressionVisitor.cs:
```diff
@@ -23,34 +23,58 @@ public class SearchConditionConvertingExpressionVisitor : SqlExpressionVisitor
+        // !(a != b) -> (a == b)
+        private SqlExpression SimplifyNegatedBinary(SqlExpression sqlExpression)
+            => sqlExpression is SqlUnaryExpression sqlUnaryExpression
+                && sqlUnaryExpression.OperatorType == ExpressionType.Not
```
IsLogicalNot

smitpatel commented on test/EFCore.SqlServer.FunctionalTests/Query/FunkyDataQuerySqlServerTest.cs:
```diff
@@ -45,15 +45,15 @@ public override async Task String_contains_on_argument_with_wildcard_constant(bo
                 @"SELECT [f].[FirstName]
 FROM [FunkyCustomers] AS [f]
-WHERE CHARINDEX(N'%B%a%r', [f].[FirstName]) <= 0",
+WHERE NOT (CHARINDEX(N'%B%a%r', [f].[FirstName]) > 0)",
```
by design reverted optimization?

maumar replied:
yes, the optimization is not valid in general for this case - it will go back to the optimized form once we know the function metadata and can determine that the charindex(firstName) is never nullable

smitpatel commented on test/EFCore.SqlServer.FunctionalTests/Query/NorthwindAggregateOperatorsQuerySqlServerTest.cs:
```diff
@@ -1171,10 +1171,8 @@ public override void Contains_over_entityType_with_null_should_rewrite_to_identi
-SELECT CASE
-    WHEN @__entity_equality_p_0_OrderID IN (
+                @"SELECT CASE
+    WHEN NULL IN (
```
Null in subquery is never null if inner is non-nullable

smitpatel replied:
Null in subquery is always null even if it matches is null in (NULL) which returns null.

---
In reply to: [363009999](https://github.com/aspnet/EntityFrameworkCore/pull/19168#discussion_r363009999) [](ancestors = 363009999)

maumar replied:
filed: https://github.com/aspnet/EntityFrameworkCore/issues/19499

smitpatel commented on test/EFCore.SqlServer.FunctionalTests/Query/NorthwindFunctionsQuerySqlServerTest.cs:
```diff
@@ -1182,7 +1182,7 @@ public override async Task Indexof_with_emptystring(bool async)
             AssertSql(
                 @"SELECT CASE
-    WHEN N'' = N'' THEN 0
+    WHEN 1 = 1 THEN 0
```
Optimize

maumar replied:
already tracked by #19295

smitpatel commented on test/EFCore.SqlServer.FunctionalTests/Query/NorthwindGroupByQuerySqlServerTest.cs:
```diff
@@ -1690,7 +1690,7 @@ public override async Task All_after_GroupBy_aggregate2(bool async)
         FROM [Orders] AS [o]
         GROUP BY [o].[CustomerID]
-        HAVING SUM([o].[OrderID]) < 0) THEN CAST(1 AS bit)
+        HAVING NOT (SUM([o].[OrderID]) >= 0)) THEN CAST(1 AS bit)
```
Why Not leaves space after it!!! :volcano:

maumar replied:
you wrote that code :trollface:

maumar commented:
PR feedback addressed @smitpatel

smitpatel commented on src/EFCore.Relational/Query/RelationalParameterBasedQueryTranslationPostprocessor.cs:
```diff
@@ -38,308 +24,32 @@ public class RelationalParameterBasedQueryTranslationPostprocessor
             Check.NotNull(parametersValues, nameof(parametersValues));
 
             var canCache = true;
+            var (sqlExpressionOptimized, optimizerCanCache) = new SqlExpressionOptimizingExpressionVisitor(
```
NullSemantics in name

maumar replied:
NullabilityBasedSqlProcessingExpressionVisitor

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1543 @@
+            var newElseResult = (SqlExpression)VisitInternal(caseExpression.ElseResult);
+            Nullable |= nullable;
+            Optimize = optimize;
+            RestoreNonNullableColumnsList(currentNonNullableColumnsCount);
```
VisitInternal will take care of resetting this

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1543 @@
+        {
+            Check.NotNull(fromSqlExpression, nameof(fromSqlExpression));
+
+            return fromSqlExpression;
```
return Check.NotNUll

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1543 @@
+
+                var processedValuesExpression = (SqlConstantExpression)VisitInternal(SqlExpressionFactory.Constant(inValues, typeMapping));
+
+                return (processedValuesExpression, (List<object>)processedValuesExpression.Value, hasNullValue);
```
just use inValues

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1543 @@
+                    inValues.Add(value);
+                }
+
+                var processedValuesExpression = (SqlConstantExpression)VisitInternal(SqlExpressionFactory.Constant(inValues, typeMapping));
```
do you really need to visit this?

maumar replied:
nope, removing

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1543 @@
+            {
+                case ExpressionType.Equal:
+                {
+                    var left = (SqlExpression)VisitInternal(predicate.Left, optimize: true);
```
should rename to allowOptimizedExpansion. optimize seems misleading

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1543 @@
+        {
+            var match = (SqlExpression)VisitInternal(likeExpression.Match);
+            var matchNullable = Nullable;
+            var pattern = (SqlExpression)VisitInternal(likeExpression.Pattern);
```
Try making visit internal generic on T so you can avoid this casting

maumar replied:
i was debating it, doesn't change much IMO but I can add it, sure

maumar replied:
i was debating it, doesn't change much IMO but I can add it, sure

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1543 @@
+        {
+            Check.NotNull(rowNumberExpression, nameof(rowNumberExpression));
+
+            var optimize = Optimize;
```
>optimize [](start = 16, length = 8)

VisitInternal will preserve optimize, no?

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1543 @@
+            var rightNullable = Nullable;
+            if (sqlBinaryExpression.OperatorType == ExpressionType.OrElse)
+            {
+                var rightNonNullableColumns = NonNullableColumns.Skip(currentNonNullableColumnsCount).ToList();
```
>rightNonNullableColumns [](start = 20, length = 23)

Inline variable and avoid ToList allocation

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1543 @@
+                RestoreNonNullableColumnsList(currentNonNullableColumnsCount);
+            }
+
+            if (sqlBinaryExpression.OperatorType == ExpressionType.Coalesce)
```
never true

smitpatel commented on src/EFCore.Relational/Query/SqlExpressionOptimizingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1543 @@
+        protected virtual IReadOnlyDictionary<string, object> ParameterValues { get; }
+        protected virtual List<ColumnExpression> NonNullableColumns { get; } = new List<ColumnExpression>();
+
+        protected virtual bool Nullable { get; set; }
```
discussed encapsulation of nullabe/optimize & nonnullable columns

smitpatel commented on src/EFCore.SqlServer/Query/Internal/SearchConditionConvertingExpressionVisitor.cs:
```diff
@@ -23,34 +23,58 @@ public class SearchConditionConvertingExpressionVisitor : SqlExpressionVisitor
+        private SqlExpression SimplifyNegatedBinary(SqlExpression sqlExpression)
+            => sqlExpression is SqlUnaryExpression sqlUnaryExpression
+                && sqlUnaryExpression.IsLogicalNot()
+                && (sqlUnaryExpression.Type == typeof(bool) || sqlUnaryExpression.Type == typeof(bool?))
```
IsLogicalNot already checks for type. Remove.

smitpatel commented on src/EFCore.Relational/Query/NullabilityBasedSqlProcessingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1493 @@
+        /// </summary>
+        /// <typeparam name="TResult">Type of the resulting expression. </typeparam>
+        /// <param name="expression">Expression that is to be visited. </param>
+        /// <param name="allowOptimizedExpansion">True if null semantics inside the expression can be expanded in the optimized way (i.e. <see langword="null"/> and <see langword="false"/> are interchangable), false otherwise. </param>
```
@AndriySvyryd @bricelam - guidance on usage of langword? or c? or nothing for null?

maumar replied:
removed the langword as we don't seem to be using it anywhere else

smitpatel commented on src/EFCore.Relational/Query/NullabilityBasedSqlProcessingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1493 @@
+            bool restoreNonNullableColumnInformation = true)
+            where TResult : Expression
+        {
+            var currentNonNullableColumnsCount = NonNullableColumns.Count;
```
short circuit for null expression before setting up all the states.

smitpatel commented on src/EFCore.Relational/Query/NullabilityBasedSqlProcessingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1493 @@
+            var nullable = caseExpression.ElseResult == null;
+            var currentNonNullableColumnsCount = NonNullableColumns.Count;
+
+            var operand = VisitInternal<SqlExpression>(caseExpression.Operand).ResultExpression;
```
Do we have a test where operand is nullable?

maumar replied:
yup, SpatialQuerySqlServerGeographyTest.OgcGeometryType

smitpatel commented on src/EFCore.Relational/Query/NullabilityBasedSqlProcessingExpressionVisitor.cs:
```diff
@@ -0,0 +1,1493 @@
+
+            var leftNonNullableColumns = NonNullableColumns.Skip(currentNonNullableColumnsCount).ToList();
+
+            _nullable = false;
```
VisitInternal on right resets this.

maumar commented:
@smitpatel another update
